src/main/scala/org/apache/spark/deploy/armada/submit/ArmadaClientApplication.scala

```diff
@@ -264,73 +509,147 @@ private[spark] class ArmadaClientApplication extends SparkApplication {
+    val (templateAnnotations, templateLabels) = armadaJobConfig.jobTemplate
+      .map { templateRequest =>
+        if (templateRequest.jobRequestItems.nonEmpty) {
+          val firstItem = templateRequest.jobRequestItems.head
+          (firstItem.annotations, firstItem.labels)
+        } else {
+          (Map.empty[String, String], Map.empty[String, String])
+        }
+      }
+      .getOrElse((Map.empty[String, String], Map.empty[String, String]))
+
+    // Resolve configuration values
+    val resolvedConfig = resolveConfig(armadaJobConfig.cliConfig, conf)
+
     val configGenerator = new ConfigGenerator("armada-spark-config", conf)
-    val annotations = configGenerator.getAnnotations ++ armadaJobConfig.nodeUniformityLabel
-      .map(label =>
-        GangSchedulingAnnotations(
-          None,
-          1 + executorCount,
-          label
+    val annotations = configGenerator.getAnnotations ++
+      templateAnnotations ++
+      armadaJobConfig.cliConfig.nodeUniformityLabel
+        .map(label =>
+          GangSchedulingAnnotations(
+            None,
+            1 + executorCount,
+            label
+          )
         )
-      )
-      .getOrElse(Map.empty)
-    val globalLabels = armadaJobConfig.podLabels
-    val driverLabels =
-      globalLabels ++ armadaJobConfig.driverLabels
-    val nodeSelectors = armadaJobConfig.nodeSelectors
+        .getOrElse(Map.empty)
+
+    // Merge template labels with config labels, config takes precedence
+    val globalLabels  = templateLabels ++ armadaJobConfig.cliConfig.podLabels
+    val driverLabels  = globalLabels ++ armadaJobConfig.cliConfig.driverLabels
+    val nodeSelectors = armadaJobConfig.cliConfig.nodeSelectors
 
     val driverPort = 7078
 
     val confSeq = conf.getAll.flatMap { case (k, v) =>
       Seq("--conf", s"$k=$v")
     }
-    val driver = newSparkDriverJobSubmitRequestItem(
-      armadaJobConfig.armadaClusterUrl,
-      armadaJobConfig.namespace,
-      armadaJobConfig.priority,
-      annotations,
-      driverLabels,
-      armadaJobConfig.containerImage,
-      driverPort,
-      clientArguments.mainClass,
-      configGenerator.getVolumes,
-      configGenerator.getVolumeMounts,
-      nodeSelectors,
-      confSeq ++ primaryResource ++ clientArguments.driverArgs,
-      conf
-    )
+    val driver = armadaJobConfig.driverJobItemTemplate match {
+      case Some(template) =>
+        // Use driver template and merge with runtime configuration
+        mergeDriverTemplate(
+          template,
+          resolvedConfig,
+          armadaJobConfig,
+          annotations,
+          driverLabels,
+          driverPort,
+          clientArguments.mainClass,
+          configGenerator.getVolumes,
+          configGenerator.getVolumeMounts,
+          nodeSelectors,
+          confSeq ++ primaryResource ++ clientArguments.driverArgs,
+          conf
+        )
+      case None =>
+        // Use default driver creation
+        newSparkDriverJobSubmitRequestItem(
```

Comment:
Why is this called: "newSparkDriverJobSubmitRequestItem"  and the executor version called: "newExecutorJobSubmitItem"?  Would it be clearer if they were: "newDriverJobSubmitRequestItem" and "newExecutorJobSubmitRequestItem"

Reply:
or maybe "newDriverJobItem"